### PR TITLE
Allow for proper garbage collection by pooling bytes.Buffer

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Allow for proper garbage collection by pooling bytes.Buffer

## Motivation and Context
Reduce memory usage by allowing quick GC sweeps.

## How Has This Been Tested?
Manually

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by poornas** [comment]
gobEncode signature has changed,tests need to be updated

File: cmd/rpc/server.go [hunk 1]
```diff
@@ -242,20 +238,18 @@ func (server *Server) ServeHTTP(w http.ResponseWriter, req *http.Request) {
...
 
-	data, err := gobEncode(callResponse)
-	if err != nil {
-		w.WriteHeader(http.StatusInternalServerError)
-		return
-	}
+	gob.NewEncoder(w).Encode(callResponse)
 
-	w.Write(data)
+	w.(http.Flusher).Flush()
```
Comment (poornas) [inline cmd/rpc/server.go:252]: shouldn't we make sure here that the writer supports flushing?

Comment (harshavardhana) [reply cmd/rpc/server.go:252]: HTTP response writer always supports flushing we don't need to validate if it supports or not.

File: cmd/rpc/client.go [hunk 2]
```diff
@@ -48,22 +47,25 @@ func (client *Client) Call(serviceMethod string, args, reply interface{}) error
 		return fmt.Errorf("rpc reply must be a pointer type, but found %v", replyKind)
 	}
 
-	data, err := gobEncode(args)
-	if err != nil {
+	b1 := bufPool.Get()
```
Comment (krisis) [inline cmd/rpc/client.go]: Let's use meaningful names like `argsBuf` and `reqBuf` instead of `b1` and `b2`.

File: cmd/rpc/pool.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+		New: func() interface{} {
+			return &bytes.Buffer{}
+		},
+	}}
+}
+
+// Get retrieves a bytes.Buffer from the pool, creating one if necessary.
+func (p Pool) Get() *bytes.Buffer {
+	buf := p.p.Get().(*bytes.Buffer)
+	buf.Reset()
```
Comment (krisis) [inline cmd/rpc/pool.go]: `buf.Reset` has already been called in `Pool.Put`. We don't need it here too.

**Review by krisis — DISMISSED** [review]

File: cmd/rpc/pool.go (hunk 3 above)
Comment (harshavardhana) [reply cmd/rpc/pool.go]: Doesn't hurt to do it again.

Comment (krisis) [reply cmd/rpc/pool.go]: What is the reason behind calling `buf.Reset` in both `Get` and `Put`?

Comment (harshavardhana) [reply cmd/rpc/pool.go]: The reason is since Get() and Put() are not synchronized, it is to avoid any unexpected situations.

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :x: [more...](#c4e62e8d-4c4c-4a29-a8ba-16fb0904cbf3) |
<a name="c4e62e8d-4c4c-4a29-a8ba-16fb0904cbf3"></a>
### 6266-53ca74e/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30985
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 679e02584346:/mint/log /tmp/mint-logs'
(1/13) Running awscli tests ... FAILED in 19 seconds
{
  "name": "awscli",
  "duration": 8084,
  "function": "make_bucket\n",
  "status": "FAIL",
  "error": "('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',))"
}

Executed 0 out of 13 tests successfully.
```

**Review by krisis — APPROVED** [review]

**Review by poornas — APPROVED** [review]

**Pull request merged**